### ISSUE TITLE
Cannot find contrastive loss in your repository

### ISSUE THREAD

**Issue by MSH970515** [jiaqihuang01/DETRIS#11]
Hello, Thanks for great work.

I cannot find the text-to-visual contrastive loss you have mentioned in the paper in the code.

The only loss implemented is binary cross entropy with logits loss, am I going right?

Can you provide the text-to-visual contrastive loss implementation?

Thank you

**Comment by jiaqihuang01** [comment]
Thank you for your question and for your interest in our work. The text-to-visual contrastive loss we mentioned in the paper is indeed implemented, but it might not be explicitly visible in the code as a separate loss function. Instead, we followed the approach from the CRIS work, which you can find more details about here: [GitHub Issue](https://github.com/DerrickWang005/CRIS.pytorch/issues/5#issuecomment-1522810584).

If you have any further questions or need clarification, feel free to ask!

**Comment by MSH970515** [comment]
Thank you for your reply. I checked the GitHub Issue link you sincerely provided.

However, I still cannot understand about the clear concept using BCE as contrastive loss.

I am assuming that it is not standard contrastive loss, but utilizing BCE as somewhat contrastive thing, as the text-guided logits are trained to be similar to GT mask.

Could you provide more detailed explanation? 

Thank you.

**Comment by kkakkkka** [comment]
Hello,

Thank you for your follow-up and for looking into the details.

To clarify: yes, you are absolutely correct that the code uses `binary_cross_entropy_with_logits` as the loss function. There is no separate `ContrastiveLoss` class (like InfoNCE) explicitly defined.

However, in the context of our paper (and following the methodology of prior works like CRIS), we describe this as a **text-to-visual contrastive loss** because of the **mechanism** by which the prediction is generated and optimized.

Here is the detailed explanation of the concept:

1. **Prediction as Similarity:**
In the `forward` method of `segmenter.py`, the final prediction is generated by the `Projector`:
```python
pred = self.proj(fq, state)

```


The `Projector` takes both the dense visual features (`fq`) and the global text embedding (`state`) to compute a score map. This score map `pred` essentially represents the **similarity** (or alignment) between the text query and each pixel in the image.
2. **BCE as Dense Contrastive Learning:**
While `binary_cross_entropy_with_logits` is a standard classification loss, when applied to this similarity map, it functions as a **dense pixel-wise contrastive objective**:
* **Positive Pairs:** For pixels inside the ground truth mask (`mask=1`), the loss forces the model to maximize the similarity between the text and the visual features (pulling them together).
* **Negative Pairs:** For pixels in the background (`mask=0`), the loss forces the model to minimize the similarity (pushing them apart).


3. **Why we call it "Contrastive":**
Unlike standard segmentation (which classifies pixels based on fixed class weights), our model classifies pixels based on their **contrast** with the specific input text. The BCE loss explicitly supervises this contrast: it trains the visual features of the referent to align with the text embedding, while contrasting them against the background.

In summary, the "text-to-visual contrastive loss" mentioned in the paper refers to the **objective** of aligning text and pixel embeddings using BCE as the optimization proxy.

Best regards.
